src/fetch_cnpj_info.py

```diff
@@ -0,0 +1,114 @@
+from concurrent import futures
+import json
+import numpy as np
+import os.path
+import pandas as pd
+import pickle
+import shutil
+from urllib.error import HTTPError
+from urllib.request import urlopen
+
+INFO_DATASET_PATH = 'data/cnpj_info.xz'
+TEMP_PATH = 'data/cnpj_info'
+
+def load_info_dataset():
+    if os.path.exists(INFO_DATASET_PATH):
+        return pd.read_csv(INFO_DATASET_PATH)
+    else:
+        return pd.DataFrame(columns=['atividade_principal',
+                                     'data_situacao',
+                                     'tipo',
+                                     'nome',
+                                     'telefone',
+                                     'atividades_secundarias',
+                                     'situacao',
+                                     'bairro',
+                                     'logradouro',
+                                     'numero',
+                                     'cep',
+                                     'municipio',
+                                     'uf',
+                                     'abertura',
+                                     'natureza_juridica',
+                                     'fantasia',
+                                     'cnpj',
+                                     'ultima_atualizacao',
+                                     'status',
+                                     'complemento',
+                                     'email',
+                                     'efr',
+                                     'motivo_situacao',
+                                     'situacao_especial',
+                                     'data_situacao_especial'])
+
+def read_csv(name):
+    return pd.read_csv('data/2016-08-08-%s.xz' % name,
+                       parse_dates=[16],
+                       dtype={'document_id': np.str,
+                              'congressperson_id': np.str,
+                              'congressperson_document': np.str,
+                              'term_id': np.str,
+                              'cnpj_cpf': np.str,
+                              'reimbursement_number': np.str})
+
+def remaining_cnpjs(info_dataset):
+    datasets = [read_csv(name)
+                for name in ['current-year', 'last-year', 'previous-years']]
+    dataset = pd.concat(datasets)
+    del(datasets)
+    is_cnpj = dataset['cnpj_cpf'].str.len() == 14.
+    cnpj_list = set(dataset.loc[is_cnpj, 'cnpj_cpf'])
+    already_fetched = set(info_dataset['cnpj'].str.replace(r'[./-]', ''))
+    return list(cnpj_list - already_fetched)
```

Comment:
Sure thing! Sorry about that 💦

A function/method that uses `yield` instead of `return` is a *generator*. This is specially useful when it comes to memory usage. So let's go back to the diference between `list`, `set`, `tuple` (all of them) and *generators*.

When you build a `list` (or `tuple` or `set`) all its elements are calculated and are available. If you type `[0, 1, 2, 3]` this is actually stored in memory, all the four values.

However, if you type `range(4)` what is created is a *generator*, an instruction, an iterator that can be used to build a `list`, `tuple`, `set` or just used as any other iterator. It tends to use small amounts of memory than the full list, but when you iterate it, the result is the same:

```python
for n in [0, 1, 2, 3]:
    print(n)
```

Outputs the same thing as:

```python
for n in range(4):
    print(n)
```

The difference is how the iterator is created. The first example is a full list stored in memory, while the second is an instructor that calculates the [`next`](https://docs.python.org/3.5/library/functions.html#next) element **only** when (and **if** needed). Yet when the *next* value arrives, the previous one is thrown away (the previous value is not in memory anymore).

In simple examples like that the difference is not meaningful at all. But if you try to use large lists, or lists with items that are large themselves, you can feel the difference. Try these examples: the first uses a list comprehension to crate an actual `list` from a *generator*; the second example is just a generator. Try them:

```python
my_iter = [number in range(6 * 10 ** 23)]
for n in my_iter:  
    print(n)
```

```python
my_iter = range(6 * 10 ** 23)
for n in my_iter:
    print(n)
```

The first code might crash. The second one runs smoothly.

So, what I'm actually suggesting is a smart memory management: getting one CNPJ at time, just when needed. And as I commented that might be useful **if** this is a huge list.

Does it make any sense?